src/credentials/credential-wallet.ts

```diff
@@ -411,8 +421,49 @@ export class CredentialWallet implements ICredentialWallet {
   /**
    * {@inheritDoc ICredentialWallet.findByQuery}
    */
-  async findByQuery(query: ProofQuery): Promise<W3CCredential[]> {
-    return this._storage.credential.findCredentialsByQuery(query);
+  async findByQuery(query: ProofQuery, opts?: { autoRefresh?: boolean }): Promise<W3CCredential[]> {
+    let creds = await this._storage.credential.findCredentialsByQuery(query);
+
+    if (!opts?.autoRefresh) {
+      return creds;
+    }
+
+    if (!this._refreshService) {
+      throw new Error('please provide credential refresh service');
+    }
+
+    let skippedCredSubjectFilter;
+    // if no creds found, let's skip credentialSubject filter and try to refresh expired creds
+    if (!creds.length && query.credentialSubject) {
+      skippedCredSubjectFilter = query.credentialSubject;
+      query.credentialSubject = undefined;
+      creds = await this._storage.credential.findCredentialsByQuery(query);
+    }
+
+    // all expired creds
+    const expiredCreds = creds.filter(
+      (c) =>
+        c.expirationDate &&
+        new Date(c.expirationDate) < new Date() &&
+        c.refreshService &&
+        c.refreshService.type === RefreshServiceType.Iden3RefreshService2023
+    );
+
+    // refresh and update storage
+    for (let i = 0; i < expiredCreds.length; i++) {
```

Comment:
added promises for fetching refreshed cred,
for delete/save, I guess better to keep it in sequence, in case some saving throws some error and not all creds saved but all removed